tools/scripts/rebuildAnimationLibraryManifest.rb

```diff
@@ -0,0 +1,205 @@
+
+class ManifestBuilder
+  def initialize(options)
+    @options = options
+  end
+
+  #
+  # Main Entry Point
+  # See end of file for possible options, parsed from command line
+  #
+  def rebuild_animation_library_manifest
+    animations_by_name = {}
+
+    # Connect to S3 and get a listing of all objects in the animation library bucket
+    info 'Reading object list from S3...'
+    bucket = Aws::S3::Bucket.new(DEFAULT_S3_BUCKET)
+    bucket.objects.each do |object_summary|
+      animation_name = object_summary.key[/^[^.]+/]
+      extension = object_summary.key[/(?<=\.)\w+$/]
+      verbose <<-EOS.unindent
+        #{bold object_summary.key}
+          #{object_summary.last_modified} | #{object_summary.size}
+      EOS
+      # Push into animations collection
+      animations_by_name[animation_name] ||= {}
+      animations_by_name[animation_name][extension] = object_summary
+    end
+
+    # TODO: Validate that keys are unique
+
+    info "Found #{animations_by_name.size} animations."
+
+    # Parallelize metadata construction because some objects will require an
+    # extra S3 request to get version IDs or image dimensions.
+    info "Building animation metadata..."
+    metadata_progress_bar = ProgressBar.create(total: animations_by_name.size) unless @options[:verbose] || @options[:quiet]
+    animation_metadata_by_name = Hash[Parallel.map(animations_by_name) do |name, objects|
+      # TODO: Validate that every JSON is paired with a PNG and vice-versa
+      # Actually download the JSON from S3
+      json_response = objects['json'].get
+      metadata = JSON.parse(json_response.body.read)
+      # TODO: Validate metadata, ensure it has everything it needs
+      # Record target version in the metadata, so environments (and projects)
+      # consistently reference the version they originally imported.
+      metadata['version'] = objects['png'].object.version_id
+
+      # Generate appropriate sourceUrl pointing to the animation library API
+      metadata['sourceUrl'] = "/api/v1/animation-library/#{metadata['version']}/#{name}.png"
+
+      # Populate sourceSize if not already present
+      unless metadata.key?('sourceSize')
+        png_body = objects['png'].object.get.body.read
+        metadata['sourceSize'] = dimensions_from_png(png_body)
+      end
+
+      verbose <<-EOS
+#{bold name} @ #{metadata['version']}
+#{JSON.pretty_generate metadata}
+      EOS
+
+      metadata_progress_bar.increment unless @options[:verbose] || @options[:quiet]
+      [name, metadata]
+    end]
+    metadata_progress_bar.finish unless @options[:verbose] || @options[:quiet]
+
+    info "Metadata built for #{animation_metadata_by_name.size} animations."
+
+    info "Building alias map..."
+    alias_progress_bar = ProgressBar.create(total: animation_metadata_by_name.size) unless @options[:quiet]
+    alias_map = {}
+    animation_metadata_by_name.each do |name, metadata|
+      aliases = [name] + (metadata['aliases'] || [])
```

Comment:
The or'ing with `[]` is in case `metadata['aliases']` is `nil`?

If so, consider the more readable (at least to me)
```
aliases = [name]
aliases += metadata['aliases'] unless metadata['aliases'].nil?
```